main/streams/streams.c

```diff
@@ -1089,37 +1089,39 @@ static size_t _php_stream_write_buffer(php_stream *stream, const char *buf, size
 
 
 	while (count > 0) {
-		towrite = count;
+		size_t towrite = count;
 		if (towrite > stream->chunk_size)
 			towrite = stream->chunk_size;
 
 		justwrote = stream->ops->write(stream, buf, towrite);
+		if (justwrote < 0) {
```

Comment:
I'm somewhat unsure about how to best handle this code. The problem is that we are performing multiple write operations here, and it could happen that the first few work fine and we write N bytes, and later an operation fails. In that case, I'm currently reporting an error, but this loses the information about the N successfully written bytes. Maybe instead I should return N with the expectation that the caller will repeat the write call and get an error reported at that point?

I'm not sure why this code actually exists though ... chunking only really makes sense for read operations, not writes. Though even if I drop this weird chunking code, the above question would still apply to buffered reads (if we want to propagate errors there as well). In that case immediately reporting the error might actually lose buffered data, which seems very bad. So I think the right thing to do here is that if at least one write succeeded, we should be reporting the successfully written bytes.

Reply:
The write chunking was introduced in https://github.com/php/php-src/commit/258aa4d239398226073d98100ba227343221c59a with the comment "Should resolve problems with network writes". Not sure whether that's still relevant and whether the "problem" was someone not running fwrite in a loop.

Reply:
Now reporting the successfully written bytes.